Source/Core/Core/PowerPC/JitArm64/JitArm64_Integer.cpp

```diff
@@ -1359,16 +1382,68 @@ void JitArm64::srawx(UGeckoInstruction inst)
       ComputeRC0(gpr.GetImm(a));
     return;
   }
-
-  if (gpr.IsImm(b) && !js.op->wantsCA)
+  else if (gpr.IsImm(s) && gpr.GetImm(s) == 0)
+  {
+    gpr.SetImmediate(a, 0);
+    ComputeCarry(false);
+    if (inst.Rc)
+      ComputeRC0(0);
+    return;
+  }
+  else if (gpr.IsImm(b))
   {
     int amount = gpr.GetImm(b);
-    if (amount & 0x20)
-      amount = 0x1F;
+
+    bool special = amount & 0x20;
+    amount &= 0x1f;
+
+    if (special)
+    {
+      gpr.BindToRegister(a, a == s);
+
+      ASR(gpr.R(a), gpr.R(s), 31);
+
+      if (js.op->wantsCA)
+      {
+        CMN(gpr.R(s), gpr.R(s));
+        ComputeCarry();
+      }
+    }
+    else if (amount == 0)
+    {
+      if (a != s)
+      {
+        gpr.BindToRegister(a, false);
+
+        ADDS(gpr.R(a), gpr.R(s), 0);
+        ComputeCarry();
+      }
+      else
+      {
+        ComputeCarry(false);
+      }
+    }
+    else if (!js.op->wantsCA)
+    {
+      gpr.BindToRegister(a, a == s);
+
+      ASR(gpr.R(a), gpr.R(s), amount);
+    }
     else
-      amount &= 0x1F;
-    gpr.BindToRegister(a, a == s);
-    ASR(gpr.R(a), gpr.R(s), amount);
+    {
+      gpr.BindToRegister(a, a == s);
+
+      ARM64Reg WA = gpr.GetReg();
+
+      LSL(WA, gpr.R(s), 32 - amount);
+      ASR(gpr.R(a), gpr.R(s), amount);
+      TST(WA, gpr.R(a));
+      CSET(WA, CC_NEQ);
```

Comment:
After thinking 10 minutes about this algorithm, I think we should have a small comment why this actually is the valid result:
LSL shifts the hidden bits to the MSB (like a rotation) with zero in the LSBs
ASR duplicates the sign bit
bitwise AND of both yields if any high bit was shifted out of the range && the sign bit was set.

IMO a very clever approach, and I assume we should use so in the lower code as well (the lower code was generated with gcc....)

However this code here can be optimized quite a bit:
TST (alias for ANDS) is a shifted register instruction. higher cost, however it includes a LSL on the second operand for free. Use it.

It hurts quite a bit to emit CSET/CMP just to move the carry from Z to C, but I have not found a better approach :(

Reply:
Untested, but feel free to pick it: https://gist.github.com/degasus/d837423ad814ff741802243e9520ecd2

Reply:
The shifted variant of TST is one cycle slower than the normal variant (at least on the CPUs I checked), so I don't think this approach is worth it if we have to use a MOV. But I've made us use it when `a != s`.